dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/filter-fee-record-payment-entity-groups.ts

```diff
@@ -0,0 +1,86 @@
+import { Currency, PremiumPaymentsFilters } from '@ukef/dtfs2-common';
+import { FeeRecordPaymentEntityGroup } from '../../../../../types/fee-record-payment-entity-group';
+import { ValidatedPaymentDetailsFilters } from '../../../../../types/utilisation-reports';
+
+/**
+ * All possible filters for FeeRecordPaymentGroup filtering.
+ */
+type Filters = PremiumPaymentsFilters & ValidatedPaymentDetailsFilters;
+
+/**
+ * Represents a function that filters FeeRecordPaymentEntityGroup objects.
+ * The Filterer should return true if the group satisfies the filter the Filterer is implementing.
+ * @param group The FeeRecordPaymentEntityGroup to be filtered.
+ * @returns A boolean indicating whether the group satisfies the filter.
+ */
+type Filterer = (group: FeeRecordPaymentEntityGroup) => boolean;
+
+/**
+ * Creates a filterer function for filtering FeeRecordPaymentEntityGroups based on facility ID.
+ * @param facilityIdFilter - The facility ID to filter by. Partial matches are allowed.
+ * @returns A Filterer function that returns true if any fee record in the group has a facility ID that includes the filter string.
+ */
+const getFacilityIdFilterer = (facilityIdFilter: string): Filterer => {
+  return (group: FeeRecordPaymentEntityGroup) => group.feeRecords.some(({ facilityId }) => facilityId.includes(facilityIdFilter));
+};
+
+/**
+ * Creates a filterer function for filtering FeeRecordPaymentEntityGroups based on payment currency.
+ * @param paymentCurrencyFilter - The payment currency to filter by. Exact matches are required.
+ * @returns A Filterer function that returns true if any payment in the group has a currency that exactly matches the filter.
+ */
+const getPaymentCurrencyFilterer = (paymentCurrencyFilter: Currency): Filterer => {
+  return (group: FeeRecordPaymentEntityGroup) => group.payments.some(({ currency }) => currency === paymentCurrencyFilter);
+};
+
+/**
+ * Creates a filterer function for filtering FeeRecordPaymentEntityGroups based on payment reference.
+ * @param paymentReferenceFilter - The payment reference to filter by. Partial matches are allowed.
+ * @returns A Filterer function that returns true if any payment in the group has a reference that includes the filter string.
+ */
+const getPaymentReferenceFilterer = (paymentReferenceFilter: string): Filterer => {
```

Comment:
have these functions been tested themselves?

Reply:
These are tested implicitly through the tests for `filterFeeRecordPaymentEntityGroups`, happy to export these 3 filterers for testing if you'd prefer?

Reply:
yes please, all functions should be texted.  There are a some other files in this PR which need the same

Reply:
Added tests for these, and for the two get functions in `dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/get-utilisation-report-reconciliation-details.test.ts`